### PR TITLE
dash/20290-custom-component-add

### LINKED ISSUE #20290: Show custom components in "Add component" list
# Description of the feature
[Docs reference](https://www.highcharts.com/docs/dashboards/edit-mode#:~:text=tablets%20and%20smartphones.-,Add%20Component,-The%20Add%20Component)

Make user defined components visible in the "Add component" list

# Library related to the feature
Dashboards

----

**You can vote for this feature by adding a thumbs-up reaction to this post.**

### PR DESCRIPTION
**Pull request by karolkolodziej**
Added a way to drag custom components from the `sidebar`, closes #20290.

### CODE REVIEW

File: ts/Dashboards/Plugins/KPIComponent.ts [hunk 1]
```diff
@@ -632,6 +634,28 @@ class KPIComponent extends Component {
...
+                    ...options,
+                    connector: {
+                        id: connectorsIds[0]
+                    }
+                };
+            }
+
+            return sidebar.onDropNewComponent(dropContext, options);
+        }
+    }
```
Comment (sebastianbochan) [inline ts/Dashboards/Plugins/KPIComponent.ts:657]: The body of the function is quite similar to other modules. I suggest creating a general function and making `options` as params. Then call it in the `onDrop` component.

File: ts/Dashboards/Actions/DragDrop.ts [hunk 2]
```diff
@@ -595,7 +595,7 @@ class DragDrop {
         mouseRowContext: Row
     ): void {
         const dragDrop = this,
-            dropPointerSize = dragDrop.options.dropPointerSize,
+            dropPointerSize = dragDrop.options.dropPointerSize || 0,
```
Comment (sebastianbochan) [inline ts/Dashboards/Actions/DragDrop.ts:598]: What is the reason of `0` placed in many params? The `dragDrop.options.dropPointerSize` can be declared in defaultOptions or replaced with no-optional variable.

File: ts/Dashboards/EditMode/SidebarPopup.ts [hunk 3]
```diff
@@ -48,168 +45,62 @@ const {
...
-                                id: connectorsIds[0]
-                            }
-                        };
+            Bindings.addComponent({
+                type: 'HTML',
+                cell: cellName,
+                elements: [
+                    {
+                        tagName: 'div',
+                        style: { 'text-align': 'center' },
```
Comment (sebastianbochan) [inline ts/Dashboards/EditMode/SidebarPopup.ts:95]: Please remove inline styling from the code. Use CSS instead.

**Review by sebastianbochan — CHANGES_REQUESTED** [review]

File: ts/Dashboards/Plugins/KPIComponent.ts (hunk 1 above)
Comment (karolkolodziej) [reply ts/Dashboards/Plugins/KPIComponent.ts:657]: All the parameters here, among them `dropPointerSize`, should be optional. The default value is declared, the `|| 0`is only to trick the TS here 😉

File: ts/Dashboards/Actions/DragDrop.ts (hunk 2 above)
Comment (karolkolodziej) [reply ts/Dashboards/Actions/DragDrop.ts:598]: All the parameters here, among them `dropPointerSize`, should be optional. The default value is declared, the `|| 0`is only to trick the TS here 😉

**Review by sebastianbochan — APPROVED** [review]

**Review by Vazonik — APPROVED** [review]

**Pull request merged**